Add Jump Consistent Hashing

Jahrme opened the pull request:
@smburdick Let me know if there is anything else you would like added.

Commit:
C implementation of Jump Consistent Hashing

Commit:
Improved Testing and Structure

Commit:
Ignore bin/ directories

Commit:
Add newline to end of Makefile

Commit:
Be consistent with int type

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+    double mean = (double) num_keys / (double) num_buckets;
+
+    double variance = 0.0;
+    for ( int i = 0 ; i < num_buckets ; ++i )
```
`for (int ...` with or without space. Pick one.

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+    int value = rand();
+
+    for (int key = 0; key < num_keys; ++key)
+    {
```
brace should be on same line. fix everywhere.

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+#include <time.h>
+#include <math.h>
+
+#include "../src/jump.c"
```
Make a header file `jump.h` and include the function prototype. Do this for `jch_test` (call it `jump_test`) also

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+    /* Start keys at random value. */
+    srand(time(0));
+
+    for (int key = 0; key < num_keys; ++key)
```
define `key` outside loop

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+    printf("Dividing %i random keys between %i buckets.\n\n", num_keys, num_buckets);
+
+    int bucket_items[num_buckets];
+    memset(bucket_items, 0, num_buckets*sizeof(int));
```
`num_buckets * sizeof(int)`

smburdick replied:
fix everywhere

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+    /* Test hashing sequential data. */
+    int exit_sequential = test_sequential(num_keys, num_buckets);
+
+    return exit_random & exit_sequential;
```
Why `&`... it's not clear to me why you're returning the values you are, so you should document this. Really what you should be doing is having the tests check against expected values and returning whether the output matches that. Since you're doing random tests that might not be feasible, so please add justification for why you're doing what you are.

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+
+int statistics(int num_keys, int num_buckets, int bucket_items[])
+{
+    double mean = (double) num_keys / (double) num_buckets;
```
casting twice is unnecessary

smburdick commented on jump/src/jump.c:
```diff
@@ -0,0 +1,21 @@
+    int64_t b = -1;
+    int64_t j = 0;
+
+    double divisor;
```
move these to same line. I think declaring `b` and `j` on separate lines is acceptable, though

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+
+int test_random(int num_keys, int num_buckets);
+int test_sequential(int num_keys, int num_buckets);
+int statistics(int num_keys, int num_buckets, int bucket_items[]);
```
put these prototypes in the header so overall test suite can call them

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+
+int main(int argc, char* argv[])
+{
+    /* Default Values */
```
I personally prefer `// ` for 1-line comments as it's easier to type but I think this way is prettier and some people insist on doing it this way. what do you think

smburdick commented on jump/src/jump.c:
```diff
@@ -0,0 +1,21 @@
+
+static const uint64_t jch_seed = 2862933555777941757;
+
+int32_t JumpConsistentHash (uint64_t key, int32_t num_buckets)
```
C functions are generally named like `jump_consistent_hash` which is how you named your tests

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+    return statistics(num_keys, num_buckets, bucket_items);
+}
+
+int statistics(int num_keys, int num_buckets, int bucket_items[])
```
name is too generic

smburdick commented on jump/tst/jch_test.c:
```diff
@@ -0,0 +1,94 @@
+    return exit_random & exit_sequential;
+}
+
+int test_random(int num_keys, int num_buckets)
```
name alone doesn't give me a good enough idea of what you're testing. What is random about it? A block comment before it would be helpful

smburdick replied:
applies for all functions

smburdick commented on jump/Makefile:
```diff
@@ -0,0 +1,33 @@
+	@echo "Cleaning..."
+	@if test -d $(BIN); then rm -rf $(BIN)/*; else mkdir $(BIN); fi
+
+compile:
```
This isn't how C makefiles generally work. what we do is we put each output file in a separate rule so we don't have to recompile everything every time we `make`. I know I skimped on this in `Bitmap-Engine` but that's a separate mess to deal with... from now on, just do it this way: https://www.gnu.org/software/make/manual/make.html#Rule-Example

smburdick reviewed (changes requested):
- Looks fine for a first C outing, but there's a few things that need to be addressed
- Names in general are too generic and comments are skimpy
- You need some test functions that entail adding/removing buckets

Commit:
Updated Makefile

Commit:
Update Consistent Hash Implementation

Commit:
Updated Testing

Commit:
Fix imports

Commit:
Link Math Library

Jahrme commented:
I believe I have addressed all of your comments. I also confirmed that this runs correctly on the server (which led to fac06d2 as on the server it does not automatically link the math library and requires the `-lm` flag to work).

smburdick commented on jump/tst/jump_test.h:
```diff
@@ -0,0 +1,30 @@
+
+uint64_t long_random ();
+
+void run_distrobution_tests
```
`run_distribution_tests`
also why the indent for the parameters
